File: CircularDoublyLinkedList.java
import java.util.Scanner;

public class CircularDoublyLinkedList {
    
    // Node sebagai static inner class
    static class Node {
        String berita;
        Node next;
        Node prev;
        
        Node(String berita) {
            this.berita = berita;
            this.next = null;
            this.prev = null;
        }
    }
    
    static Node head = null;
    static Node tail = null;
    static int size = 0;
    
    static void insertBerita(String berita) {
        Node newNode = new Node(berita);
        
        if (head == null) {
            head = newNode;
            tail = newNode;
            head.next = head;
            head.prev = head;
        } else {
            tail.next = newNode;
            newNode.prev = tail;
            newNode.next = head;
            head.prev = newNode;
            tail = newNode;
        }
        size++;
        System.out.println("Berita ditambahkan");
    }
    
    static boolean hapusBerita(int nomor) {
        if (head == null || nomor < 1 || nomor > size) {
            return false;
        }
        
        Node current = head;
        for (int i = 1; i < nomor; i++) {
            current = current.next;
        }
        
        if (size == 1) {
            head = null;
            tail = null;
        } else {
            current.prev.next = current.next;
            current.next.prev = current.prev;
            
            if (current == head) {
                head = current.next;
            }
            if (current == tail) {
                tail = current.prev;
            }
        }
        size--;
        return true;
    }
    
    static void tampilForward() throws InterruptedException {
        if (head == null) {
            System.out.println("Tidak ada berita");
            return;
        }
        
        System.out.println("\nBERITA FORWARD:");
        Node current = head;
        int nomor = 1;
        do {
            System.out.println(nomor + ". " + current.berita);
            current = current.next;
            nomor++;
            Thread.sleep(3000);
        } while (current != head);
    }
    
    static void tampilBackward() throws InterruptedException {
        if (head == null) {
            System.out.println("Tidak ada berita");
            return;
        }
        
        System.out.println("\nBERITA BACKWARD:");
        Node current = tail;
        int nomor = size;
        do {
            System.out.println(nomor + ". " + current.berita);
            current = current.prev;
            nomor--;
            Thread.sleep(3000);
        } while (current != tail);
    }
    
    static void tampilBerita(int nomor) {
        if (head == null || nomor < 1 || nomor > size) {
            System.out.println("Nomor tidak valid");
            return;
        }
        
        Node current = head;
        for (int i = 1; i < nomor; i++) {
            current = current.next;
        }
        System.out.println("Berita " + nomor + ": " + current.berita);
    }
    
    static void tampilSemua() {
        if (head == null) {
            System.out.println("Tidak ada berita");
            return;
        }
        
        System.out.println("\nDAFTAR BERITA:");
        Node current = head;
        int nomor = 1;
        do {
            System.out.println(nomor + ". " + current.berita);
            current = current.next;
            nomor++;
        } while (current != head);
    }
    
    public static void main(String[] args) throws InterruptedException {
        Scanner input = new Scanner(System.in);
        int pilih;
        
        do {
            System.out.println("\nMENU:");
            System.out.println("1. Insert berita");
            System.out.println("2. Hapus berita");
            System.out.println("3. Tampilkan berita secara forward");
            System.out.println("4. Tampilkan berita secara backward");
            System.out.println("5. Tampilkan berita tertentu");
            System.out.println("6. Exit");
            System.out.print("Pilih: ");
            pilih = input.nextInt();
            input.nextLine();
            
            switch(pilih) {
                case 1:
                    System.out.print("Masukkan teks berita: ");
                    String berita = input.nextLine();
                    insertBerita(berita);
                    break;
                    
                case 2:
                    tampilSemua();
                    if (head != null) {
                        System.out.print("Masukkan nomor berita yang akan dihapus: ");
                        int nomorHapus = input.nextInt();
                        if (hapusBerita(nomorHapus)) {
                            System.out.println("Berita nomor " + nomorHapus + " berhasil dihapus");
                        } else {
                            System.out.println("Gagal menghapus");
                        }
                    }
                    break;
                    
                case 3:
                    tampilForward();
                    break;
                    
                case 4:
                    tampilBackward();
                    break;
                    
                case 5:
                    tampilSemua();
                    if (head != null) {
                        System.out.print("Masukkan nomor berita yang ingin ditampilkan: ");
                        int nomorTampil = input.nextInt();
                        tampilBerita(nomorTampil);
                    }
                    break;
                    
                case 6:
                    System.out.println("Program selesai");
                    break;
                    
                default:
                    System.out.println("Pilihan tidak valid");
            }
        } while (pilih != 6);
        
        input.close();
    }
}

File: CircularDoublyLinkedList.py
import time

class Node:
    def __init__(self, berita):
        self.berita = berita
        self.next = None
        self.prev = None

head = None
tail = None
size = 0

def insertBerita(berita):
    global head, tail, size
    newNode = Node(berita)
    
    if head is None:
        head = newNode
        tail = newNode
        head.next = head
        head.prev = head
    else:
        tail.next = newNode
        newNode.prev = tail
        newNode.next = head
        head.prev = newNode
        tail = newNode
    size += 1
    print("Berita ditambahkan")

def hapusBerita(nomor):
    global head, tail, size
    if head is None or nomor < 1 or nomor > size:
        return False
    
    current = head
    for i in range(1, nomor):
        current = current.next
    
    if size == 1:
        head = None
        tail = None
    else:
        current.prev.next = current.next
        current.next.prev = current.prev
        
        if current == head:
            head = current.next
        if current == tail:
            tail = current.prev
    size -= 1
    return True

def tampilForward():
    if head is None:
        print("Tidak ada berita")
        return
    
    print("\nBERITA FORWARD:")
    current = head
    nomor = 1
    while True:
        print(f"{nomor}. {current.berita}")
        current = current.next
        nomor += 1
        time.sleep(3)
        if current == head:
            break

def tampilBackward():
    if head is None:
        print("Tidak ada berita")
        return
    
    print("\nBERITA BACKWARD:")
    current = tail
    nomor = size
    while True:
        print(f"{nomor}. {current.berita}")
        current = current.prev
        nomor -= 1
        time.sleep(3)
        if current == tail:
            break

def tampilBerita(nomor):
    if head is None or nomor < 1 or nomor > size:
        print("Nomor tidak valid")
        return
    
    current = head
    for i in range(1, nomor):
        current = current.next
    print(f"Berita {nomor}: {current.berita}")

def tampilSemua():
    if head is None:
        print("Tidak ada berita")
        return
    
    print("\nDAFTAR BERITA:")
    current = head
    nomor = 1
    while True:
        print(f"{nomor}. {current.berita}")
        current = current.next
        nomor += 1
        if current == head:
            break

def main():
    while True:
        print("\nMENU:")
        print("1. Insert berita")
        print("2. Hapus berita")
        print("3. Tampilkan berita secara forward")
        print("4. Tampilkan berita secara backward")
        print("5. Tampil berita tertentu")
        print("6. Exit")
        
        try:
            pilih = int(input("Pilih: "))
        except:
            print("Input angka")
            continue
        
        if pilih == 1:
            berita = input("Berita: ")
            insertBerita(berita)
        
        elif pilih == 2:
            tampilSemua()
            if head is not None:
                try:
                    nomor = int(input("Nomor yang dihapus: "))
                    if hapusBerita(nomor):
                        print(f"Berita {nomor} dihapus")
                    else:
                        print("Gagal menghapus")
                except:
                    print("Input salah")
        
        elif pilih == 3:
            tampilForward()
        
        elif pilih == 4:
            tampilBackward()
        
        elif pilih == 5:
            tampilSemua()
            if head is not None:
                try:
                    nomor = int(input("Nomor yang ditampil: "))
                    tampilBerita(nomor)
                except:
                    print("Input salah")
        
        elif pilih == 6:
            print("Selesai")
            break
        
        else:
            print("Pilihan salah")

if __name__ == "__main__":
    main()